File: results/0.5/GZIP/gzip-meta-llama-llama-3.3-70b-instruct-turbo.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdint.h>
#include <string.h>
#include <stdlib.h>

#define ID1 0x1F
#define ID2 0x8B
#define CM 8

typedef struct {
    uint8_t id1;
    uint8_t id2;
    uint8_t cm;
    uint8_t flags;
    uint32_t mtime;
    uint8_t xfl;
    uint8_t os;
} gzip_header_t;

typedef struct {
    uint16_t xlen;
    uint8_t* extra;
} gzip_extra_t;

typedef struct {
    uint8_t* fname;
    uint8_t* fcomment;
} gzip_names_t;

typedef struct {
    uint16_t fhcrc;
} gzip_fhcrc_t;

typedef struct {
    uint8_t* compr_data;
    uint32_t isize;
} gzip_compr_data_t;

typedef struct hammer_parser {
    uint8_t* data;
    size_t size;
    size_t pos;
    int error;
    char* error_message;
} hammer_parser_t;

hammer_parser_t* hammer_new_parser(uint8_t* data, size_t size) {
    hammer_parser_t* p = malloc(sizeof(hammer_parser_t));
    p->data = data;
    p->size = size;
    p->pos = 0;
    p->error = 0;
    p->error_message = NULL;
    return p;
}

void hammer_free_parser(hammer_parser_t* p) {
    free(p);
}

uint8_t hammer_get_byte(hammer_parser_t* p) {
    if (p->pos >= p->size) {
        p->error = 1;
        p->error_message = "End of data reached";
        return 0;
    }
    return p->data[p->pos++];
}

uint16_t hammer_get_uint16_le(hammer_parser_t* p) {
    if (p->pos + 2 > p->size) {
        p->error = 1;
        p->error_message = "End of data reached";
        return 0;
    }
    uint16_t val = p->data[p->pos] | (p->data[p->pos + 1] << 8);
    p->pos += 2;
    return val;
}

uint32_t hammer_get_uint32_le(hammer_parser_t* p) {
    if (p->pos + 4 > p->size) {
        p->error = 1;
        p->error_message = "End of data reached";
        return 0;
    }
    uint32_t val = p->data[p->pos] | (p->data[p->pos + 1] << 8) | (p->data[p->pos + 2] << 16) | (p->data[p->pos + 3] << 24);
    p->pos += 4;
    return val;
}

uint8_t* hammer_get_bytes(hammer_parser_t* p, size_t len) {
    if (p->pos + len > p->size) {
        p->error = 1;
        p->error_message = "End of data reached";
        return NULL;
    }
    uint8_t* bytes = malloc(len);
    memcpy(bytes, p->data + p->pos, len);
    p->pos += len;
    return bytes;
}

uint8_t* hammer_get_null_terminated_string(hammer_parser_t* p) {
    size_t len = 0;
    while (p->pos + len < p->size && p->data[p->pos + len] != 0) {
        len++;
    }
    if (p->pos + len >= p->size) {
        p->error = 1;
        p->error_message = "End of data reached";
        return NULL;
    }
    uint8_t* str = malloc(len + 1);
    memcpy(str, p->data + p->pos, len);
    str[len] = 0;
    p->pos += len + 1;
    return str;
}

uint8_t* hammer_get_remaining_bytes(hammer_parser_t* p) {
    size_t len = p->size - p->pos;
    uint8_t* bytes = malloc(len);
    memcpy(bytes, p->data + p->pos, len);
    p->pos = p->size;
    return bytes;
}

hammer_parser_t* gzip_header(hammer_parser_t* p) {
    uint8_t id1 = hammer_get_byte(p);
    if (id1 != ID1) {
        p->error = 1;
        p->error_message = "Invalid ID1";
        return p;
    }
    uint8_t id2 = hammer_get_byte(p);
    if (id2 != ID2) {
        p->error = 1;
        p->error_message = "Invalid ID2";
        return p;
    }
    uint8_t cm = hammer_get_byte(p);
    if (cm != CM) {
        p->error = 1;
        p->error_message = "Invalid compression method";
        return p;
    }
    return p;
}

hammer_parser_t* gzip_flags(hammer_parser_t* p, uint8_t* flags) {
    *flags = hammer_get_byte(p);
    return p;
}

hammer_parser_t* gzip_mtime(hammer_parser_t* p, uint32_t* mtime) {
    *mtime = hammer_get_uint32_le(p);
    return p;
}

hammer_parser_t* gzip_xfl(hammer_parser_t* p, uint8_t* xfl) {
    *xfl = hammer_get_byte(p);
    return p;
}

hammer_parser_t* gzip_os(hammer_parser_t* p, uint8_t* os) {
    *os = hammer_get_byte(p);
    return p;
}

hammer_parser_t* gzip_extra(hammer_parser_t* p, gzip_extra_t* extra) {
    extra->xlen = hammer_get_uint16_le(p);
    extra->extra = hammer_get_bytes(p, extra->xlen);
    return p;
}

hammer_parser_t* gzip_fname(hammer_parser_t* p, uint8_t** fname) {
    *fname = hammer_get_null_terminated_string(p);
    return p;
}

hammer_parser_t* gzip_fcomment(hammer_parser_t* p, uint8_t** fcomment) {
    *fcomment = hammer_get_null_terminated_string(p);
    return p;
}

hammer_parser_t* gzip_fhcrc(hammer_parser_t* p, gzip_fhcrc_t* fhcrc) {
    fhcrc->fhcrc = hammer_get_uint16_le(p);
    return p;
}

hammer_parser_t* gzip_compr_data(hammer_parser_t* p, gzip_compr_data_t* compr_data) {
    compr_data->compr_data = hammer_get_remaining_bytes(p);
    return p;
}

hammer_parser_t* gzip_isize(hammer_parser_t* p, uint32_t* isize) {
    *isize = hammer_get_uint32_le(p);
    return p;
}

hammer_parser_t* gzip_parser(hammer_parser_t* p) {
    gzip_header_t header;
    gzip_extra_t extra;
    gzip_names_t names;
    gzip_fhcrc_t fhcrc;
    gzip_compr_data_t compr_data;

    p = gzip_header(p);
    p = gzip_flags(p, &header.flags);
    p = gzip_mtime(p, &header.mtime);
    p = gzip_xfl(p, &header.xfl);
    p = gzip_os(p, &header.os);

    if (header.flags & 0x04) {
        p = gzip_extra(p, &extra);
    }

    if (header.flags & 0x08) {
        p = gzip_fname(p, &names.fname);
    }

    if (header.flags & 0x10) {
        p = gzip_fcomment(p, &names.fcomment);
    }

    if (header.flags & 0x02) {
        p = gzip_fhcrc(p, &fhcrc);
    }

    p = gzip_compr_data(p, &compr_data);
    p = gzip_isize(p, &compr_data.isize);

    return p;
}

int main(int argc, char** argv) {
    if (argc != 2) {
        printf("Usage: %s <input_file>\n", argv[0]);
        return 1;
    }

    FILE* file = fopen(argv[1], "rb");
    if (!file) {
        printf("Error opening file %s\n", argv[1]);
        return 1;
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    rewind(file);

    uint8_t* data = malloc(file_size);
    fread(data, 1, file_size, file);
    fclose(file);

    hammer_parser_t* p = hammer_new_parser(data, file_size);
    p = gzip_parser(p);

    if (p->error) {
        printf("Error parsing gzip file: %s\n", p->error_message);
        return 1;
    }

    free(data);
    hammer_free_parser(p);

    return 0;
}

File: results/0.5/GZIP/gzip-meta-llama-llama-3.3-70b-instruct-turbo.rs
use nom::{
    bytes::complete::{tag, take},
    error::{ErrorKind, Err, ParseError},
    number::complete::{be_u16, be_u32, be_u8},
    IResult,
};
use std::env;
use std::fs::File;
use std::io::{Read, Result};

#[derive(Debug)]
struct GzipHeader {
    id1: u8,
    id2: u8,
    cm: u8,
    flags: u8,
    mtime: u32,
    xfl: u8,
    os: u8,
    extra: Option<Vec<u8>>,
    fname: Option<String>,
    fcomment: Option<String>,
    hcrc: Option<u16>,
}

fn parse_gzip_header(input: &[u8]) -> IResult<&[u8], GzipHeader> {
    let (input, _) = tag([0x1f, 0x8b])(input)?;
    let (input, cm) = be_u8(input)?;
    let (input, flags) = be_u8(input)?;
    let (input, mtime) = be_u32(input)?;
    let (input, xfl) = be_u8(input)?;
    let (input, os) = be_u8(input)?;

    let (input, extra) = if flags & 0x04 != 0 {
        let (input, xlen) = be_u16(input)?;
        let (input, extra) = take(xlen)(input)?;
        (input, Some(extra.to_vec()))
    } else {
        (input, None)
    };

    let (input, fname) = if flags & 0x08 != 0 {
        let (input, fname) = take_while(|c| c != 0)(input)?;
        (input, Some(String::from_utf8_lossy(fname).into_owned()))
    } else {
        (input, None)
    };

    let (input, fcomment) = if flags & 0x10 != 0 {
        let (input, fcomment) = take_while(|c| c != 0)(input)?;
        (input, Some(String::from_utf8_lossy(fcomment).into_owned()))
    } else {
        (input, None)
    };

    let (input, hcrc) = if flags & 0x02 != 0 {
        let (input, hcrc) = be_u16(input)?;
        (input, Some(hcrc))
    } else {
        (input, None)
    };

    Ok((
        input,
        GzipHeader {
            id1: 0x1f,
            id2: 0x8b,
            cm,
            flags,
            mtime,
            xfl,
            os,
            extra,
            fname,
            fcomment,
            hcrc,
        },
    ))
}

fn take_while<F>(predicate: F) -> impl Fn(&[u8]) -> IResult<&[u8], &[u8]>
where
    F: Fn(u8) -> bool + Clone,
{
    move |input: &[u8]| {
        let mut i = 0;
        while i < input.len() && predicate(input[i]) {
            i += 1;
        }
        if i == 0 {
            return Err(Err::Error(nom::error::Error::new(input, ErrorKind::NonEmpty)));
        }
        Ok((&input[i..], &input[..i]))
    }
}

fn main() -> Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        println!("Usage: {} <input_file>", args[0]);
        return Ok(());
    }

    let mut file = File::open(&args[1])?;
    let mut data = Vec::new();
    file.read_to_end(&mut data)?;

    match parse_gzip_header(&data) {
        Ok((remaining, header)) => {
            println!("Gzip Header: {:?}", header);
            println!("Remaining data: {:?}", remaining);
        }
        Err(err) => {
            println!("Error parsing gzip header: {:?}", err);
        }
    }

    Ok(())
}